src/coreclr/vm/frozenobjectheap.cpp

```diff
@@ -135,34 +157,59 @@ FrozenObjectSegment::FrozenObjectSegment(size_t sizeHint) :
         ThrowOutOfMemory();
     }
 
+    m_pStart = static_cast<uint8_t*>(committedAlloc);
+    m_pCurrent = m_pStart + sizeof(ObjHeader);
+    m_SizeCommitted = FOH_COMMIT_SIZE;
+
     // ClrVirtualAlloc is expected to be PageSize-aligned so we can expect
     // DATA_ALIGNMENT alignment as well
     _ASSERT(IS_ALIGNED(committedAlloc, DATA_ALIGNMENT));
+}
 
-    segment_info si;
-    si.pvMem = committedAlloc;
-    si.ibFirstObject = sizeof(ObjHeader);
-    si.ibAllocated = si.ibFirstObject;
-    si.ibCommit = FOH_COMMIT_SIZE;
-    si.ibReserved = m_Size;
-
-    m_SegmentHandle = GCHeapUtilities::GetGCHeap()->RegisterFrozenSegment(&si);
-    if (m_SegmentHandle == nullptr)
+void FrozenObjectSegment::RegisterOrUpdate(uint8_t* current, size_t sizeCommited)
+{
+    CONTRACTL
     {
-        ClrVirtualFree(alloc, 0, MEM_RELEASE);
-        ThrowOutOfMemory();
+        THROWS;
+        MODE_PREEMPTIVE;
     }
+    CONTRACTL_END
 
-    m_pStart = static_cast<uint8_t*>(committedAlloc);
-    m_pCurrent = m_pStart + sizeof(ObjHeader);
-    m_SizeCommitted = si.ibCommit;
-    INDEBUG(m_ObjectsCount = 0);
-    return;
+    if (VolatileLoad(&m_pCurrentRegistered) == nullptr)
+    {
+        segment_info si;
+        si.pvMem = m_pStart;
+        si.ibFirstObject = sizeof(ObjHeader);
+        si.ibAllocated = (size_t)m_pCurrentRegistered;
+        si.ibCommit = sizeCommited;
+        si.ibReserved = m_Size;
+
+        // NOTE: RegisterFrozenSegment may take a GC lock inside.
+        m_SegmentHandle = GCHeapUtilities::GetGCHeap()->RegisterFrozenSegment(&si);
+        if (m_SegmentHandle == nullptr)
+        {
+            ThrowOutOfMemory();
+        }
+        VolatileStore(&m_pCurrentRegistered, current);
+    }
+    else
+    {
+        if (current > VolatileLoad(&m_pCurrentRegistered))
+        {
+            GCHeapUtilities::GetGCHeap()->UpdateFrozenSegment(
+                m_SegmentHandle, current, m_pStart + sizeCommited);
+            VolatileStore(&m_pCurrentRegistered, current);
+        }
+        else
+        {
+            // Some other thread already advanced it.
+        }
```

Comment:
```suggestion
    if (m_pCurrentRegistered == nullptr)
    {
        segment_info si;
        si.pvMem = m_pStart;
        si.ibFirstObject = sizeof(ObjHeader);
        si.ibAllocated = (size_t)current;
        si.ibCommit = sizeCommited;
        si.ibReserved = m_Size;

        // NOTE: RegisterFrozenSegment may take a GC lock inside.
        m_SegmentHandle = GCHeapUtilities::GetGCHeap()->RegisterFrozenSegment(&si);
        if (m_SegmentHandle == nullptr)
        {
            ThrowOutOfMemory();
        }
        m_pCurrentRegistered = current;
    }
    else
    {
        if (current > m_pCurrentRegistered)
        {
            GCHeapUtilities::GetGCHeap()->UpdateFrozenSegment(
                m_SegmentHandle, current, m_pStart + sizeCommited);
            m_pCurrentRegistered = current;
        }
        else
        {
            // Some other thread already advanced it.
        }
```
The registration used `m_pCurrentRegistered` that would always be null.
Also, `m_pCurrentRegistered` is always read under the `m_SegmentRegistrationCrst` lock. It does not need to volatile.